sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/meta/provider/iceberg/IcebergReadWriteIT.java

```diff
@@ -0,0 +1,174 @@
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Integration tests for writing to Iceberg with Beam SQL. */
+@RunWith(JUnit4.class)
+public class IcebergReadWriteIT {
+  private static final Schema SOURCE_SCHEMA =
+      Schema.builder()
+          .addNullableField("c_bigint", INT64)
+          .addNullableField("c_integer", INT32)
+          .addNullableField("c_float", FLOAT)
+          .addNullableField("c_double", DOUBLE)
+          .addNullableField("c_boolean", BOOLEAN)
+          .addNullableField("c_timestamp", CalciteUtils.TIMESTAMP)
+          .addNullableField("c_varchar", STRING)
+          .addNullableField("c_char", STRING)
+          .addNullableField("c_arr", Schema.FieldType.array(STRING))
+          .build();
+
+  @Rule public transient TestPipeline writePipeline = TestPipeline.create();
+  @Rule public transient TestPipeline readPipeline = TestPipeline.create();
+
+  private static final BigqueryClient BQ_CLIENT = new BigqueryClient("IcebergReadWriteIT");
+  static final String DATASET = "iceberg_sql_tests_" + System.nanoTime();
+  static String warehouse;
+  protected static final GcpOptions OPTIONS =
+      TestPipeline.testingPipelineOptions().as(GcpOptions.class);
+
+  @BeforeClass
+  public static void createDataset() throws IOException, InterruptedException {
+    warehouse =
+        String.format(
+            "%s%s/%s",
+            TestPipeline.testingPipelineOptions().getTempLocation(),
+            IcebergReadWriteIT.class.getSimpleName(),
+            UUID.randomUUID());
+    BQ_CLIENT.createNewDataset(OPTIONS.getProject(), DATASET);
+  }
+
+  @AfterClass
+  public static void deleteDataset() {
+    BQ_CLIENT.deleteDataset(OPTIONS.getProject(), DATASET);
+  }
+
+  @Test
+  public void testSqlWriteAndRead() throws IOException, InterruptedException {
+    BeamSqlEnv sqlEnv =
+        BeamSqlEnv.inMemory(
+            IcebergTableProvider.create()
+                .withCatalogProperties(
+                    ImmutableMap.of(
+                        "catalog-impl", "org.apache.iceberg.gcp.bigquery.BigQueryMetastoreCatalog",
+                        "io-impl", "org.apache.iceberg.gcp.gcs.GCSFileIO",
+                        "warehouse", warehouse,
+                        "gcp_project", OPTIONS.getProject(),
+                        "gcp_region", "us-central1")));
+    String tableIdentifier = DATASET + ".my_table";
+
+    // 1) create beam table
+    String createTableStatement =
+        "CREATE EXTERNAL TABLE TEST( \n"
+            + "   c_bigint BIGINT, \n"
+            + "   c_integer INTEGER, \n"
+            + "   c_float FLOAT, \n"
+            + "   c_double DOUBLE, \n"
+            + "   c_boolean BOOLEAN, \n"
+            + "   c_timestamp TIMESTAMP, \n"
+            + "   c_varchar VARCHAR, \n "
+            + "   c_char CHAR, \n"
+            + "   c_arr ARRAY<VARCHAR> \n"
```

Comment:
Could you also add a nested array field such as array<array<String>> and  some nested types also ?

Reply:
Yep will do

Reply:
Thanks for suggesting this, it pointed out some incorrect conversion logic: #34887

Reply:
Fixed conversion logic and push a new commit

(lmk if you'd like me to separate it into a different PR)